### PR TITLE
Upgrade build scripts to use latest SDK/NDK and latest Gradle

### LINKED ISSUE #446: Android build is broken with latest SDK/NDK/Gradle and JDK
Clean checkout doesn't build out of the box on latest version of the tools required. I am in ArchLinux and all my tools are up to date but current Android build system gives random interesting errors.

### PR DESCRIPTION
**Pull request by wasimabbas-arm**
## Description

This PR upgrades the build system to work with latest SDK, NDK, Gradle and JDK. 
Fixes #446 

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

N/A

### CODE REVIEW

File: bldsys/cmake/create_gradle_project.cmake
Comment (gpx1000) [inline bldsys/cmake/create_gradle_project.cmake]: Is it required to set the ANDROID_SDK environment variable?

File: bldsys/cmake/template/gradle/local.properties.in
Comment (gpx1000) [inline bldsys/cmake/template/gradle/local.properties.in]: Isn't it bad practice to include local.properties in source repo?  I thought this file is auto generated by AS.

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (gpx1000) [inline bldsys/cmake/template/gradle/app.build.gradle.in:1]: What's the difference between the build.gradle and the app build.gradle?  I know if you include the app build.gradle one must change to using gradlew app:assembleDebug etc rather than simply gradlew assembleDebug.

**Review by gpx1000 — COMMENTED** [review]
Have a few questions but overall looks good to me.

File: bldsys/cmake/template/gradle/local.properties.in
Comment (wasimabbas-arm) [reply bldsys/cmake/template/gradle/local.properties.in]: TBH, not sure. Rememeber our instructions are for non-AS based builds. i.e. straight from command line, in which case this is not generated.

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (wasimabbas-arm) [reply bldsys/cmake/template/gradle/app.build.gradle.in:1]: `build.gradle` is suppose to have generic info in it and `app/build.gradle` will have app specific info. For example you can't have an `android{}` section in `build.gradle` anymore. You will get weird errors if you do. This is why `app/build.gradle` is introduced and `android{}` section is moved there.

> gradlew app:assembleDebug 

Yes because assembleDebug is now inside `app`. However from command line just using `gradle assembleDebug` still works fine.

File: bldsys/cmake/create_gradle_project.cmake
Comment (wasimabbas-arm) [reply bldsys/cmake/create_gradle_project.cmake]: I might have misunderstood how cmake works but its not setting 'environment variable' here. Its just a local variable that I have used inside `local.propertise`. Why I didn't just use `ANDROID_HOME` in local.propertise is because I don't know how to cmake `config file` using enviroment variables. I think it only works with local variables. Someone can correct me if I am wrong.

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (wasimabbas-arm) [reply bldsys/cmake/template/gradle/app.build.gradle.in:1]: Also this whole new structure is consistent with what latest Android Studio (Bumblebee) will generate if you create a new project.

**Comment by SaschaWillems** [comment]
Can you add the used component versions? I have updated everything in my Windows machine, and I can't build with this PR.

**Comment by SaschaWillems** [comment]
Can you specify the component versions you're used? I have updated all Android components on my Windows machine and the build fails with a syntax error in line 49 of build.gradle.

File: bldsys/cmake/create_gradle_project.cmake
Comment (gpx1000) [reply bldsys/cmake/create_gradle_project.cmake]: It is just a local variable being used here.  You'd need ENV to set it to an environment variable.  So you can do it with this: https://cmake.org/cmake/help/latest/variable/ENV.html

File: bldsys/cmake/template/gradle/local.properties.in
Comment (gpx1000) [reply bldsys/cmake/template/gradle/local.properties.in]: Right, I forgot about the instructions.  I use AS for any android building and CLion rest of the time; our instructions are less efficient than using IDEs.

Comment (gpx1000) [reply bldsys/cmake/template/gradle/local.properties.in]: Given the only thing it's being used for is the location of the SDK and duplication from the ENV variable, why not just use the ENV variable directly in CMake and call it a day?  My concern is that If someone has a custom local.properties this might interfere with it?  (I honestly don't know if it would or not, locally we ban letting local.properties into source control).

**Comment by Akbar-arm** [comment]
I am also having a similar issue on windows with line 49 of build.gradle. I get "NDK not installed" even though the latest side by side version is installed and the environment variables are set. If I set ndk.dir manually in local.properties, I start getting C/C++ Configuration problems.

**Comment by abbaswasim** [comment]
> Can you specify the component versions you're used? I have updated all Android components on my Windows machine and the build fails with a syntax error in line 49 of build.gradle.

@SaschaWillems
Sorry all, lost track of this. Not getting any updates from GitHub. 

I could never reliably tell which SDK I have but some folks say `<android sdk> \tools\source.properties` should have the version. 

The following is what I have tested with. As you can see only MacOs and ArchLinux. But @TomAtkinsonArm and @PaoloJovon-arm has tested this on Windows. Can you guys share your setup as well please?

**MacOS:** 
NDK: 23.1.7779620
SDK:  Pkg.Revision=26.1.1
JDK: 
```
java 13.0.1 2019-10-15
Java(TM) SE Runtime Environment (build 13.0.1+9)
Java HotSpot(TM) 64-Bit Server VM (build 13.0.1+9, mixed mode, sharing)
```
Gradle: 
```
------------------------------------------------------------
Gradle 7.4.2
------------------------------------------------------------

Build time:   2022-03-31 15:25:29 UTC
Revision:     540473b8118064efcc264694cbcaa4b677f61041

Kotlin:       1.5.31
Groovy:       3.0.9
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          18.0.1 (Homebrew 18.0.1+0)
OS:           Mac OS X 11.6.5 x86_64
```

Android Gradle:
```
$ ./gradlew --version

------------------------------------------------------------
Gradle 7.4.2
------------------------------------------------------------

Build time:   2022-03-31 15:25:29 UTC
Revision:     540473b8118064efcc264694cbcaa4b677f61041

Kotlin:       1.5.31
Groovy:       3.0.9
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          13.0.1 (Oracle Corporation 13.0.1+9)
OS:           Mac OS X 10.16 x86_64
```

**ArchLinux:**
NDK: 24.0.8215888
SDK:  Pkg.Revision=26.1.1
JDK: 
```
openjdk 18.0.1 2022-04-19
OpenJDK Runtime Environment (build 18.0.1+10)
OpenJDK 64-Bit Server VM (build 18.0.1+10, mixed mode)
```

Gradle: 
```
------------------------------------------------------------
Gradle 7.4.2
------------------------------------------------------------

Build time:   2022-03-31 16:40:07 UTC
Revision:     <unknown>

Kotlin:       1.5.31
Groovy:       3.0.9
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          18.0.1 (Oracle Corporation 18.0.1+10)
OS:           Linux 5.15.35-1-lts amd64
```

Android Gradle:
```
$ ./gradlew --version 
------------------------------------------------------------
Gradle 7.4.2
------------------------------------------------------------

Build time:   2022-03-31 15:25:29 UTC
Revision:     540473b8118064efcc264694cbcaa4b677f61041

Kotlin:       1.5.31
Groovy:       3.0.9
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          18.0.1 (Oracle Corporation 18.0.1+10)
OS:           Linux 5.15.35-1-lts amd64```

**Comment by Akbar-arm** [comment]
Thanks to @JoseEmilio-ARM I managed to solve my issues on windows.
Firstly, I needed to select "import project" rather than using "open" in android studio. This cleared most of the problems I was having earlier. I still occasionally get an error about C/ C++ configuration problems but the build succeeds and runs fine so I just ignore it.
I hit the syntax issue (on line 49 of build.gradle) when trying to run Build > Clean Project. Going to File > Invalidating Caches and choosing to restart android studio cleared that error too. I've never had that syntax error appear again.
image

File: bldsys/cmake/create_gradle_project.cmake
Comment (abbaswasim) [reply bldsys/cmake/create_gradle_project.cmake]: Not sure I follow. My problem is I don't know how to do a cmake `config_file` with environment variables. NOT how to read ENV var in a cmake script.

At the moment I am doing
```
sdk.dir=@ANDROID_SDK@
```
Are you saying something like the following works?
```
sdk.dir=@$ENV(ANDROID_HOME)@
```

File: bldsys/cmake/template/gradle/local.properties.in
Comment (abbaswasim) [reply bldsys/cmake/template/gradle/local.properties.in]: @gpx1000 But I don't need the sdk path in cmake. I need this in android scripts (gradle). `local.properties` would be bad if submitted in source control with hard-coded paths. In this case its only locally created to make android build happy.

File: bldsys/cmake/create_gradle_project.cmake
Comment (gpx1000) [reply bldsys/cmake/create_gradle_project.cmake]: Not quite try this:
set(ENV ANDROID_SDK "" CACHE STRING "")

Comment (gpx1000) [reply bldsys/cmake/create_gradle_project.cmake]: However, as there is a local.properties.in this is the way to do a config file:

local.properties.in:
`sdk.dir="${ANDROID_SDK}"`

create_gradle_project.cmake:
`
set(ENV ANDROID_SDK "" CACHE STRING "")
configure_file(local.properties.in local.properties)
`

So multiple ways to achieve your desired results.  If you do the configure_file version, there's no need to add to the environment vars so a simple set(ANDROID_SDK "") is enough to make things happy.

Comment (abbaswasim) [reply bldsys/cmake/create_gradle_project.cmake]: Actually `sdk.dir=$ENV{ANDROID_HOME}` would work and create a valid `local.properties` But the problem is I can't do any sanity checks on `ANDROID_HOME` in the `local.properties.in` its better done in `CMakelists.txt` as I am doing now.

File: bldsys/cmake/template/gradle/local.properties.in
Comment (gpx1000) [reply bldsys/cmake/template/gradle/local.properties.in]: Technically, we could modify it to have cmake call gradle and cmake can call build directly, but that's a bit much extra.  The configure_file option will do what you're looking for, or simply adding it to the environment variables will allow gradle to not require the local.properties I believe?

**Comment by gpx1000** [comment]
Import instead of open means that intellij (Android Studio) is generating the project files as a first step.  This shouldn't be required at all.  Open should be enough.

File: bldsys/cmake/create_gradle_project.cmake
Comment (gpx1000) [reply bldsys/cmake/create_gradle_project.cmake]: Also, can I recommend we use android.ndkPath in the projects gradle?  As of AS4.1, this would be the way to tell gradle where to find the NDK.
We could easily just place ndk.dir=blah blah blah in the local.properties.in file.  However, this is deprecated in AS4.1 (while it still works and might be cleaner as this PR adds local.properties generation at the cmake invocation.
Either way, if we define the path to the ndk, we should be able to reliably not require AndroidStudio even be opened at any point in the build process.  Also it will fix the issues reported with regards to not finding the NDK on project open.

Comment (abbaswasim) [reply bldsys/cmake/create_gradle_project.cmake]: I thought ndk has to be inside your sdk these days. This is how its finding it now. Originally I had ndk outside my sdk, multiple copies of it actually. If you setup newly these days, installing SDK with ndk component will install it inside it. So don't need to set `android.ndkPath` in the future.

File: bldsys/cmake/template/gradle/local.properties.in
Comment (abbaswasim) [reply bldsys/cmake/template/gradle/local.properties.in]: @gpx1000  It could. Having a little more thinking about this and trying to remember why I went the path route is that I think I saw somewhere some gradle messaging saying paths via environment variables is or will be deprecated. If I have go that route I am curious how do you guys setup Android SDK on your system? Which environment variable do you set?

I have quite a few and lost track of which one is for what;

```
ANDROID_NDK_ROOT=/development/android/sdk/ndk/24.0.8215888
ANDROID_SDK_ROOT=/development/android/sdk
ANDROID_HOME=/development/android/sdk
ANDROID_NDK_HOME=/development/android/sdk/ndk/24.0.8215888
ANDROID_MAKE_CCACHE=ccache
ANDROID_ROOT=/development/android/sdk
```

If I go the ENV var route then we don't need the `local.properties`. I was just under the impression the environment variables route will be deprecated soon. This is why I am reading the ENV var in CMake to pass it onto gradle as path so its happy. Gradle and AS keeps changing on day basis as well so this is painful to follow.

File: bldsys/cmake/create_gradle_project.cmake
Comment (gpx1000) [reply bldsys/cmake/create_gradle_project.cmake]: It doesn't *have* to be inside the SDK.  This is just very common.  android.version DSL is usually used to allow for gradle to do the right thing.  However, it is still a thing (I do it on CICD instances) to install the SDK without Android studio and the NDK can be installed anywhere.  It's also a setup popular in older game engine environments that haven't really been updated to the latest DSL in a while.
I usually suggest being explicit in android project creation in such a way as to where AS does the right thing whether it generates anything or not (i.e. all projects should build without AS even being installed).
One would just need the SDK+NDK.  AS is just a way to package it and make it easier to work with for developers.

**Comment by abbaswasim** [comment]
> Import instead of open means that intellij (Android Studio) is generating the project files as a first step. This shouldn't be required at all. Open should be enough.

So what's missing that its generating?

File: bldsys/cmake/template/gradle/local.properties.in
Comment (gpx1000) [reply bldsys/cmake/template/gradle/local.properties.in]: I'd be interested in what that environment vars deprecated message came from as I haven't seen it yet.

You definitely don't need both ANDROID_NDK_ROOT and ANDROID_NDK_HOME.  I see you've been developing Android a while as you have all of the old ways ;-).  So the bare minimum of what you need is ANDROID_HOME.  CCACHE can be found in other ways (i.e. if you're using GNU make then ANDROID_MAKE_CCACHE is more useful) however, in general I get CCACHE via the cmake build system identifying it and using it.

https://developer.android.com/studio/command-line/variables

I personally usually just set the ndkpath explicitly via android.ndkPath.  So on a CI environment, I'll grab the androidManager then use android.ndkPath to point at the NDK I might be using.  Which might or might not be the one installed via Android SDK Manager. (there's cases where I've had to have multiple versions installed).
https://developer.android.com/studio/projects/configure-agp-ndk

**Comment by gpx1000** [comment]
The differences between Open and Import is AS is going to lookup and create module linkages inside the .idea folder.  You will find differences in .idea/libraries/*.xml and .idea/modules.xml.  Those should only be created on Open and not Import.  Import will also generate a slightly different .iml project file which wont list libraries thus it will rely upon the system and the settings found there.
The reason it's having problems on Open is the libraries are set to look locally and not globally.  I'm currently not near a computer with AS installed and am doing this from getting chemo infused so can't test and report the exact differences between the two until I get home later and my response is from my expectations having worked on plugins for CLion (IntelliJ and AS).

**Comment by abbaswasim** [comment]
@gpx1000 removed the whole setting up of local.properties. This works for me on command line (MacOS, ArchLinux). Need some testing on Windows.

File: bldsys/cmake/create_gradle_project.cmake
Comment (abbaswasim) [reply bldsys/cmake/create_gradle_project.cmake]: I never use AS to install SDK or NDK. I always do it via the command line tools. And AFAICT this is their default behaviour now.

File: bldsys/cmake/template/gradle/local.properties.in
Comment (abbaswasim) [reply bldsys/cmake/template/gradle/local.properties.in]: ok I can verify that just leaving `ANDROID_HOME` and removing everything else including the `local.properties` still works. So I have got rid of those in the last commit. This simplifies things a bit.

Comment (abbaswasim) [reply bldsys/cmake/template/gradle/local.properties.in]: > I'd be interested in what that environment vars deprecated message came from as I haven't seen it yet.

I was thinking of this. I have seen these messages myself. https://github.com/KhronosGroup/Vulkan-Samples/pull/431

**Comment by abbaswasim** [comment]
> 

I guess what I am interested in is, what do we have to add to the build system to make it work with AS "Open" as well. Something in gradle files somewhere?

**Comment by gpx1000** [comment]
I was able to get this new version to work here, Linux, OSX, and Windows.  With these changes It LGTM;

Thanks for the work on it!

**Comment by gpx1000** [comment]
Oh, sorry didn't answer the question.  This new version should work with straight Open again.  I believe the difference is when you import IntelliJ (AS) does its best to resolve tooling so if the local system in the settings don't result in finding things, it wont list them in the .idea project settings which means it will find the system SDK and not set the local properties for gradle to look beyond the first order of preferance to resolve the SDK location.  Open doesn't do that check it just will use the local properties and report to gradle that the SDK is found in a location it isn't, gradle dutifully doesn't look further down the line as it has the settings set to tell it to look in the wrong spot.
Moving to Environment variables and reducing the local properties enables gradle to just use the system to do the right thing.

File: bldsys/cmake/template/gradle/local.properties.in
Comment (gpx1000) [reply bldsys/cmake/template/gradle/local.properties.in]: I think those messages are saying correctly that ANDROID_NDK_ROOT is deprecated in favor of ANDROID_NDK_HOME.  ANDROID_HOME is enough if the ndk is found inside the SDK.  Which is by far the most useful and correct place for it to be.

**Review by gpx1000 — DISMISSED** [review]

**Comment by SaschaWillems** [comment]
Building now works fine for me out of the box with the latest changes. Been having lots of troubles on Android with a locally installed CMake version, and that's no longer the case with this PR. So thumbs up 👍🏼

One thing I noticed though is that, at least in release mode, the generated apk now has a different name compared to before (and as stated in the readme). Instead of `vulkan_samples-release.apk` with this PR it's now named `app-release.apk`. Can that be changed back? If not the android build readme should be changed accordingly.

**Comment by abbaswasim** [comment]
@SaschaWillems [the last commit](https://github.com/KhronosGroup/Vulkan-Samples/pull/447/commits/354adad185b41d83650b1641fadf60b5e25aae11) should fix the app name issue

**Comment by SaschaWillems** [comment]
Thx. The apk is now properly named.

But I noticed that it's now put into a different folder than the one from the readme: `app\build\outputs\apk\release`

Can you check if that's the case on your side too? If so, we need to update the installation instructions in the readme.

**Comment by abbaswasim** [comment]
> Thx. The apk is now properly named.
> 
> But I noticed that it's now put into a different folder than the one from the readme: `app\build\outputs\apk\release`
> 
> Can you check if that's the case on your side too? If so, we need to update the installation instructions in the readme.

My APKs are built in the following locations:
`./build/android_gradle/app/build/outputs/apk/debug/vulkan_samples-debug.apk
./build/android_gradle/app/build/outputs/apk/release/vulkan_samples-release.apk`

The instructions says to install from:
`adb install build/outputs/apk/release/vulkan_samples-release.apk
adb install build/outputs/apk/debug/vulkan_samples-debug.apk`

@SaschaWillems  Are you saying your apks are generated in `app\build\outputs\apk\[release/debug]`? and Are you on Windows/Linux/Mac?

**Comment by SaschaWillems** [comment]
My apks are built in the same locations you posted. But they don't match the instructions from the readme. This is on Windows.

Currently, the readme tells you to install like this:
```
cd build/android_gradle
adb install build/outputs/apk/debug/vulkan_samples-debug.apk
``` 

But that fails, as the apk (with this PR, though it may have been wrong on master too) is put into
```
app\build\outputs\apk\debug
```
Relative to the `cd build/android_gradle` from the readme. So calling adb install like described in the readme, fails, as the apk is located in a different folder.

The correct command woud be
```
adb install app/build/outputs/apk/debug/vulkan_samples-debug.apk
```

But maybe I'm not reading the build instructions correct. If that's the case, just ignore my comment.

If I'm the only one having that problem, I'm fine to merge this :)

**Comment by gpx1000** [comment]
Good catch @SaschaWillems .  I think we should go ahead and change the readme.  Moving from the project gradle to the app gradle does indeed make it build under the app project.  One could conceivably equally rename app to anything and I bet it would build there.  However, if it's true that gradle is deprecating building from the main project file, then we should respect that in our documentation.

**Comment by wasimabbas-arm** [comment]
@SaschaWillems  sorry just wanted to confirm the issue. Updated the readme now. Since I have introduced the `app/build.gradle` that’s the right fix.

**Review by gpx1000 — APPROVED** [review]

**Review by SaschaWillems — APPROVED** [review]
Thanks for fixing this. With the updated readme this looks good to me :)

**Comment by TomAtkinsonArm** [comment]
@gary-sweet are you able to review this? @marty-johnson59 looking to unblock this so I can fix other PRs

**Comment by gary-sweet** [comment]
> @gary-sweet are you able to review this? @marty-johnson59 looking to unblock this so I can fix other PRs

No, sorry. I have no way to test this.

**Comment by TomAtkinsonArm** [comment]
@marty-johnson59 who would be the best person to ping for this?

CI passes, 2 reviews from vendors, lots of testing from myself, @PaoloJovon-arm and @wasimabbas-arm

**Comment by marty-johnson59** [comment]
I'll go ahead and merge as we have a couple approvals and lots of testing.  LMK if it causes issues and we can revert.

**Pull request merged**
